src/mscorlib/src/System/ArraySegment.cs

```diff
@@ -110,6 +110,87 @@ public int Count
             }
         }
 
+        public int Length => Count;
+
+        public T this[int index]
+        {
+            get
+            {
+                if (_array == null)
+                    ThrowHelper.ThrowInvalidOperationException(ExceptionResource.InvalidOperation_NullArray);
+                if (index < 0 || index >= _count)
+                    ThrowHelper.ThrowArgumentOutOfRange_IndexException();
+                Contract.EndContractBlock();
+
+                return _array[_offset + index];
+            }
+
+            set
+            {
+                if (_array == null)
+                    ThrowHelper.ThrowInvalidOperationException(ExceptionResource.InvalidOperation_NullArray);
+                if (index < 0 || index >= _count)
+                    ThrowHelper.ThrowArgumentOutOfRange_IndexException();
+                Contract.EndContractBlock();
+
+                _array[_offset + index] = value;
```

Comment:
@AlexRadch I see your point about not having to check the bounds.

I think that you might still get an overflow in `_offset + index` if both `_offset` and `index` have large values.